### ISSUE TITLE
Does Enzyme.jacobian support mutating operations as ForwardDiff.jacobian does?

### ISSUE THREAD

**Issue by junyixu** [EnzymeAD/Enzyme.jl#1370]
https://juliadiff.org/ForwardDiff.jl/v0.7/user/api.html#Jacobians-of-f(x::AbstractArray)::AbstractArray-1

An API like:
`ForwardDiff.jacobian(f!, y::AbstractArray, x::AbstractArray, cfg::JacobianConfig = JacobianConfig(f!, y, x), check=Val{true}())`
where where `f!` updates `y`.

**Comment by wsmoses** [comment]
No, but you can use the more general autodiff interface which handles arbitrary mutation/data structures.

**Comment by junyixu** [comment]
For example, 
```julia
function g(x::AbstractVector)
    return log(x[1]) + x[1]*x[2] - sin(x[2])
end

x = [2.0, 5.0]
∂g_∂x= zeros(size(x))
autodiff(Reverse, g, Active, Duplicated(x, ∂g_∂x))

# ∂g_∂x
# 2-element Vector{Float64}:
#  5.5
#  1.7163378145367738
```

but how to do the same thing with `g!`
```julia
rst_vec = [0.0]
function g!(rst_vec::AbstractVector, x::AbstractVector)
    rst_vec[1] = log(x[1]) + x[1]*x[2] - sin(x[2])
    return nothing
end
```

**Comment by junyixu** [comment]
Is there an example of mutating operations?

**Comment by junyixu** [comment]
I can give an example for forward mode.
```julia
function f!(y::Vector{Float64}, x::Vector{Float64})
    y[1] = x[1] * x[1] + x[2] * x[1]
    y[2] = x[2] * x[2] + x[1] * x[2]
    return nothing
end

x  = [2.0, 3.0]
y  = [0.0, 0.0]

dx=onehot(x)
dy=onehot(y)

Enzyme.autodiff(Forward, f!, BatchDuplicated(y, dy), BatchDuplicated(x, dx))
```
Is there one for reverse mode AD?

**Comment by gdalle** [comment]
I think if you want the specific API of ForwardDiff.jl, i.e. functions `x -> y` represented as `f!(y, x)`, you will be better served by https://github.com/gdalle/DifferentiationInterface.jl

The Enzyme bindings are probably not optimal yet (in particular, the Jacobian is not batched), but it will do what you want

**Comment by wsmoses** [comment]
@junyixu yeah we can add that if you prefer. I hadn't seen that API before so that's why we don't have it.

If you're interested in helping contribute, I'd be happy to help you with a PR.

**Comment by vchuravy** [comment]
@gdalle this should look like:
```
DifferentiationInterface.jacobian(f, y, AutoEnzyme(), x) 
```

right?

**Comment by gdalle** [comment]
Yes, and you can also specify the mode in AutoEnzyme, otherwise it will pick forward mode by default.
It's not yet super optimized and it doesn't yet use batches but at least it's there

**Comment by wsmoses** [comment]
if this is sufficiently common, we should just add this API to Enzyme.

**Comment by gdalle** [comment]
I wouldnt mind, whenever the API exists upstream I try to use it. Today I added the reverse Jacobian and the HVP that you provide (https://github.com/gdalle/DifferentiationInterface.jl/pull/445)

**Comment by wsmoses** [comment]
@gdalle do you want to open a PR with your current implementation as a starting place?

**Comment by gdalle** [comment]
The issue is that my current implementation is not specific to Enzyme: it relies heavily on the DI machinery to derive `jacobian` from `pushforward` or `pullback`. The magic happens in [this file](https://github.com/gdalle/DifferentiationInterface.jl/blob/main/DifferentiationInterface/src/first_order/jacobian.jl) but I don't think it's gonna be very heplful for a pure-Enzyme version.
